### PR TITLE
[Torch] Lower AtenLogsumexpOp via exp/sum/log decomposition

### LINKED ISSUE #4578: error: failed to legalize operation 'torch.constant.bool'
When lowering some KernelBench programs (level2: 18, 22, 42, 43, 45, 51, 58, 64), I get this error:

`error: failed to legalize operation 'torch.constant.bool'`

Reproducer:

Clone [Lighthouse](https://github.com/llvm/lighthouse)
Init uv as in README
Run: `$ uv run examples/KernelBench/test-kernel-bench.py --kernel level2/18_`

Error message:
```
error: failed to legalize operation 'torch.constant.bool'
ERROR: got an error converting /home/rengolin/devel/llvm/lighthouse/third_party/KernelBench/KernelBench/level2/18_Matmul_Sum_Max_AvgPool_LogSumExp_LogSumExp.py to MLIR:
...
  File "/home/rengolin/devel/llvm/lighthouse/.venv/lib/python3.12/site-packages/torch_mlir/compiler_utils.py", line 136, in run_pipeline_with_repro_report
    raise TorchMlirCompilerError(trimmed_message) from None
torch_mlir.compiler_utils.TorchMlirCompilerError: Lowering Torch Backend IR -> Linalg-on-Tensors Backend IR failed with the following diagnostics:


python exception: Failure while executing pass pipeline

For Torch-MLIR developers, the error can be reproduced with:
$ torch-mlir-opt -pass-pipeline='builtin.module(torch-backend-to-linalg-on-tensors-backend-pipeline{allow-non-finites=True})' /tmp/UnnammedModule.mlir
Add '-mlir-print-ir-after-all -mlir-disable-threading' to get the IR dump for debugging purpose.
```

Kernel [18](https://github.com/ScalingIntelligence/KernelBench/blob/018c5997ffc1f0fb5bc4ed15fc8b55bc5fab5a4e/KernelBench/level2/18_Matmul_Sum_Max_AvgPool_LogSumExp_LogSumExp.py) has:
`x = torch.sum(x, dim=1, keepdim=True) # (batch_size, 1)`

And the others with the same error have a similar pattern: `keepdim=True`.

### PR DESCRIPTION
**Pull request by johnny19436**
Closes #4578 

This PR fixes AtenLogsumexpOp lowering in the Torch-to-Linalg path, resolving the legalization failure in the linalg-on-tensors backend pipeline.

Updates TorchToLinalg reduction handling for logsumexp. Adds a regression test to cover conversion behavior and full backend pipeline execution.

### CODE REVIEW

**Comment by rengolin** [comment]
Pre-commit error is clang-format mostly.

File: lib/Conversion/TorchToLinalg/Reduction.cpp [hunk 1]
```diff
@@ -665,6 +669,44 @@ class ConvertReductionOp : public ConversionPattern {
...
+        linalg::YieldOp::create(builder, nestedLoc, Value{result});
+      err = !result;
+    };
+
+    Value expOp = torch_to_linalg::createElementwiseLinalgGeneric(
+        rewriter, loc, {inputTensor}, elemType, expBodyBuilder);
+    return err ? Value{} : expOp;
+  }
+
+  Value createElementwiseLogsumexpLog(Location loc, Type elemType,
```
Comment (rengolin) [inline lib/Conversion/TorchToLinalg/Reduction.cpp]: This is an identical function as the above, could it not be parametrized?

File: lib/Conversion/TorchToLinalg/Reduction.cpp [hunk 2]
```diff
@@ -769,6 +811,19 @@ class ConvertReductionOp : public ConversionPattern {
     if (failed(elemTypeCheck))
       return elemTypeCheck;
 
+    if (isa<AtenLogsumexpOp>(op) && !isa<mlir::FloatType>(elemType))
+      return rewriter.notifyMatchFailure(op, "only float types are supported");
+
+    // `aten.logsumexp` is lowered as `log(sum(exp(x), dim, keepdim))`.
+    if (isa<AtenLogsumexpOp>(op)) {
```
Comment (rengolin) [inline lib/Conversion/TorchToLinalg/Reduction.cpp]: I would just move the `FloatType` check inside this block.

File: lib/Conversion/TorchToLinalg/Reduction.cpp (hunk 1 above)
Comment (johnny19436) [reply lib/Conversion/TorchToLinalg/Reduction.cpp]: updated

File: lib/Conversion/TorchToLinalg/Reduction.cpp (hunk 2 above)
Comment (johnny19436) [reply lib/Conversion/TorchToLinalg/Reduction.cpp]: updated

**Comment by rengolin** [comment]
This looks good to me now, but I'll let the repo developers approve and merge. I have added some reviewers that have previously approved PRs in this area.

**Review by zjgarvey — COMMENTED** [review]
Thanks!

Since this op is really a composite of simpler torch ops, it makes more sense to add a decomposition pattern in `Dialect/Torch/Transforms/DecomposeComplexOps.cpp`. E.g. here is a decomposition for very similar op (logsoftmax): https://github.com/llvm/torch-mlir/blob/d3e7bdd422133b74f995b21a65ef1b6f7c79d597/lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp#L3103

Alternatively, you can try to add this op to the default decompositions at the fx graph level via:

https://github.com/llvm/torch-mlir/blob/d3e7bdd422133b74f995b21a65ef1b6f7c79d597/python/torch_mlir/extras/fx_decomp_util.py#L5

Adding the op to that list would apply the following decomposition to the torch.export output (before we even import the fx.graph to torch-mlir).

https://github.com/pytorch/pytorch/blob/7d8f21b4b802f958de912891b49b21694a5e70e0/torch/_refs/__init__.py#L900 

-------
@sahas3 is the use of `float(inf)` in that link above okay for those backends which don't allow `inf`, or should we in-house a decomposition to be able to case it out?
--------

Also, please use e2e tests whenever possible. Given the linked reference implementation, I'm not sure the current implementation in your PR is going to hold up to numerics testing.

**Comment by johnny19436** [comment]
Thanks, good point. I moved logsumexp to DecomposeComplexOps. Tests were updated and pass.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 3]
```diff
@@ -3125,6 +3172,73 @@ static Value getLogSoftmaxResult(OpTy op, PatternRewriter &rewriter) {
...
+  }
+
+  Value cstTrue = ConstantBoolOp::create(rewriter, loc, true);
+  Type xMaxType = computeReductionTypeForDimList(rewriter, op, tensorType, dims,
+                                                 /*keepDim=*/true);
+  if (!xMaxType)
+    return nullptr;
+
+  Value xMax = AtenAmaxOp::create(rewriter, loc, xMaxType, self, dim, cstTrue);
+  Value shifted = createTensorSub(rewriter, loc, tensorType, self, xMax);
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: I am not seeing the equality check for `float("inf")` as done in the PyTorch decomposition. This will likely lead to incorrect answers if input has `-inf/inf`.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 4]
```diff
@@ -3125,6 +3172,73 @@ static Value getLogSoftmaxResult(OpTy op, PatternRewriter &rewriter) {
...
+  Value shiftedExp = AtenExpOp::create(rewriter, loc, tensorType, shifted);
+
+  Value dtype = ConstantNoneOp::create(rewriter, loc);
+  Value shiftedSumExp = AtenSumDimIntListOp::create(
+      rewriter, loc, op.getType(), shiftedExp, dim, op.getKeepdim(), dtype);
+
+  Value shiftedLogSumExp =
+      AtenLogOp::create(rewriter, loc, shiftedSumExp.getType(), shiftedSumExp);
+
+  Value xMaxReduced = AtenAmaxOp::create(rewriter, loc, op.getType(), self, dim,
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Here you can use the same trick as in https://github.com/pytorch/pytorch/blob/7d8f21b4b802f958de912891b49b21694a5e70e0/torch/_refs/__init__.py#L917 to squeeze the output of `amax` with `keepdim=True` instead of computing `amax` again.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 5]
```diff
@@ -85,6 +85,53 @@ static Type computeReductionType(PatternRewriter &rewriter, Operation *op,
   return resultType;
 }
 
+// Helper function to compute the return type of the reduction function for
+// list-valued `dim`.
+static Type computeReductionTypeForDimList(PatternRewriter &rewriter,
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: A lot of overlap with existing `computeReductionType` like checks for `hasSizes`, `toPositiveDim`, `keepDim` etc. Can the existing method be enhanced to avoid duplication? Alternatively, extract out the commonalities as helpers for reuse?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 6]
```diff
@@ -85,6 +85,53 @@ static Type computeReductionType(PatternRewriter &rewriter, Operation *op,
   return resultType;
 }
 
+// Helper function to compute the return type of the reduction function for
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Can you update the PR title to [Torch] as the content is now torch level decomp?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 7]
```diff
@@ -3125,6 +3172,73 @@ static Value getLogSoftmaxResult(OpTy op, PatternRewriter &rewriter) {
...
+//  x_max_reduced = aten.amax(x, dim, keepdim)
+//  logsumexp = shifted_logsumexp + x_max_reduced
+static Value getLogsumexpResult(AtenLogsumexpOp op, PatternRewriter &rewriter) {
+  Location loc = op.getLoc();
+  Value dim = op.getDim();
+  Value self = op.getSelf();
+  BaseTensorType tensorType = cast<BaseTensorType>(self.getType());
+
+  if (!tensorType.hasDtype() || !isa<mlir::FloatType>(tensorType.getDtype())) {
+    (void)rewriter.notifyMatchFailure(op, "Only support floating type");
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]:
```suggestion
    (void)rewriter.notifyMatchFailure(op, "Only floating point types supported");
```

File: test/Dialect/Torch/decompose-complex-ops.mlir [hunk 8]
```diff
@@ -1174,3 +1174,37 @@ func.func @mean_dim_scalar_negative_dim(%arg0: !torch.vtensor<[],f32>) -> !torch
...
+
+// CHECK-LABEL: func.func @logsumexp_decompose_keepdim_false
+// CHECK-NOT: torch.aten.logsumexp
+// CHECK: torch.aten.max.dim
+// CHECK: torch.aten.exp
+// CHECK: torch.aten.sum.dim_IntList
+// CHECK: torch.aten.log
+// CHECK: torch.aten.add.Tensor
+func.func @logsumexp_decompose_keepdim_false(%arg0: !torch.vtensor<[4,4],f32>) -> !torch.vtensor<[4],f32> {
+  %dim = torch.constant.int 1
```
Comment (sahas3) [inline test/Dialect/Torch/decompose-complex-ops.mlir]: Can you add tests for multi dim reduction like `dims = [0, 1]` and all dim reduction `dims = []` and negative dim? These can probably be e2e tests. Is there a preference between LIT/e2e @zjgarvey?

**Review by sahas3 — CHANGES_REQUESTED** [review]
Thanks for the change. I've left a few suggestions based on which I think it might be better to just use PyTorch's decomp directly instead of replicating the logic here.

**Comment by sahas3** [comment]
> @sahas3 is the use of float(inf) in that link above okay for those backends which don't allow inf, or should we in-house a decomposition to be able to case it out?

I think not handling non-finites will likely need some more thought. For C++ impl, we should honor the `allowNonFinites` flag but mimicking all the pytorch decomp in C++ is not scalable IMO. So in those cases, we'll already have the non-finites baked in via the fx_import process.

File: test/Dialect/Torch/decompose-complex-ops.mlir (hunk 8 above)
Comment (zjgarvey) [reply test/Dialect/Torch/decompose-complex-ops.mlir]: Yes, for op support, e2e is a far better read than lit tests.

**Comment by johnny19436** [comment]
Thanks for the reviews. I pushed an update addressing the requested changes.
I added `aten.logsumexp` to the FX decomposition table so the FX import path uses PyTorch's decomposition directly. I also kept the C++ decomposition as a fallback for Torch IR that still contains `aten.logsumexp`, and aligned it with PyTorch's handling of `inf` by masking `abs(x_max) == float("inf")` to zero.
Other updates:
- Reuse `amax(..., keepdim=True)` with squeeze instead of recomputing `amax`.
- Share dim normalization logic with `computeReductionType`.
- Add e2e coverage for multi-dim, empty-dim, and negative-dim reductions.

Verified the new e2e cases with FX importer + linalg refbackend; all pass. Please take another look.

**Comment by sahas3** [comment]
> I added `aten.logsumexp` to the FX decomposition table so the FX import path uses PyTorch's decomposition directly. I also kept the C++ decomposition as a fallback for Torch IR that still contains `aten.logsumexp`

Thanks for the change @johnny19436, however, I don't think we need the C++ decomp at all. It is effectively deadcode for fx importer path as `aten.logsumexp` will already be decomposed by the time the C++ pass processes the torch IR. You only need to keep the `DEFAULT_DECOMPOSITIONS` change and the e2e change. Rest can be reverted. Thanks!

**Comment by johnny19436** [comment]
@sahas3 Updated, ready for review!

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 9]
```diff
@@ -651,6 +651,77 @@ def ReduceSumDimIntListElementTypeBoolModule_basic(module, tu: TestUtils):
 # ==============================================================================
 
 
+class ReduceLogSumExpDimIntListFloatModule(torch.nn.Module):
```
Comment (rengolin) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py]: nit: you put the `logexp` tests in _the middle_ of the `sum` tests. Better to block them off.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py [hunk 10]
```diff
@@ -651,6 +651,77 @@ def ReduceSumDimIntListElementTypeBoolModule_basic(module, tu: TestUtils):
...
+    module_factory=lambda: ReduceLogSumExpDimIntListEmptyDimKeepDimModule()
+)
+def ReduceLogSumExpDimIntListEmptyDimKeepDimModule_basic(module, tu: TestUtils):
+    module.forward(tu.rand(3, 4, 5))
+
+
+# ==============================================================================
+
+
+class ReduceLogSumExpDimIntListNegativeDimModule(torch.nn.Module):
```
Comment (rengolin) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py]: do we also need some integer/boolean tests here? Not sure it makes a lot of sense, but the docs don't specify

Comment (johnny19436) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/reduction.py]: sure, I'll add it.

**Review by rengolin — COMMENTED** [review]
If that was all we needed, then this is even better. @sahas3

**Comment by johnny19436** [comment]
The ONNX e2e failures are dtype mismatches, not numeric errors. PyTorch’s logsumexp promotes int/bool inputs to float32, but the ONNX import path keeps int64/bool outputs. The golden reference is float32, so ReduceLogSumExpDimIntListIntModule_basic and ReduceLogSumExpDimIntListBoolModule_basic fail on --config=onnx.

**Comment by sahas3** [comment]
> The ONNX e2e failures are dtype mismatches, not numeric errors. PyTorch’s logsumexp promotes int/bool inputs to float32, but the ONNX import path keeps int64/bool outputs. The golden reference is float32, so ReduceLogSumExpDimIntListIntModule_basic and ReduceLogSumExpDimIntListBoolModule_basic fail on --config=onnx.

These are the new test points you've added? So no need to fix ONNX path (unless you are using the ONNX path and want to fix that 😄 ) since those are pre-existing failures exposed by the new test. The CI needs to pass before the change can be merged though. So the next step is to update https://github.com/llvm/torch-mlir/blob/41bc308c539cdb42b875385189120d8e4a851978/projects/pt1/e2e_testing/xfail_sets.py#L2838 with these two expected failures.

**Review by sahas3 — APPROVED** [review]
Thanks for all the changes, LG.
Let me know if you need help with merging.
